### PR TITLE
Utilize cookies to store session variables

### LINKED ISSUE #18: Copying session variables between managed vm instances is broken
In the bookshelf app, http session variables are not copied between instances, which messes up auth.

### PR DESCRIPTION
**Pull request by shun-fan**
fixes #18

### CODE REVIEW

File: bookshelf/src/main/java/com/example/managedvms/gettingstartedjava/auth/LoginServlet.java [hunk 1]
```diff
@@ -57,12 +65,19 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp)
             SCOPE)
         .build();
     String state = new BigInteger(130, new SecureRandom()).toString(32);
-    req.getSession().setAttribute("state", state);
+    Cookie stateCookie = new Cookie("state", state);
+    stateCookie.setPath("/");
+    resp.addCookie(stateCookie);
+    // do i need this, or do
```
Comment (lesv) [inline bookshelf/src/main/java/com/example/managedvms/gettingstartedjava/auth/LoginServlet.java]: A bit more?

**Comment by lesv** [comment]
@shun-fan You might wish to read http://www.eclipse.org/jetty/documentation/current/session-management.html and see if any of it applies to us.  It seems relatively easy to configure, but it might not be useful at all.

I'll also want ludo to review this.

File: bookshelf/src/main/java/com/example/managedvms/gettingstartedjava/auth/Oauth2CallbackServlet.java [hunk 2]
```diff
@@ -56,15 +62,25 @@
   @Override
   public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException,
       ServletException {
+    Map<String, Cookie> cookieMap = new HashMap<>();
+    Cookie[] cookies = req.getCookies();
+    for (Cookie c : cookies) {
```
Comment (lesv) [inline bookshelf/src/main/java/com/example/managedvms/gettingstartedjava/auth/Oauth2CallbackServlet.java]: Do we always want to do this?   Shouldn't there be a comment to disable this in production?

File: bookshelf/src/main/java/com/example/managedvms/gettingstartedjava/auth/Oauth2CallbackServlet.java [hunk 3]
```diff
@@ -93,11 +109,10 @@ public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOExc
...
     // from this map, extract the relevant profile info and store it in the session
-    req.getSession().setAttribute("userEmail", userIdResult.get("email"));
-    req.getSession().setAttribute("userId", userIdResult.get("id"));
-    req.getSession().setAttribute("userImageUrl", userIdResult.get("picture"));
-    req.getRequestDispatcher(
-        req.getSession().getAttribute("loginDestination").toString()).forward(req, resp);
+    resp.addCookie(new Cookie("userEmail", userIdResult.get("email")));
+    resp.addCookie(new Cookie("userId", userIdResult.get("id")));
+    resp.addCookie(new Cookie("userImageUrl", userIdResult.get("picture")));
+    resp.sendRedirect(cookieMap.get("loginDestination").getValue());
```
Comment (lesv) [inline bookshelf/src/main/java/com/example/managedvms/gettingstartedjava/auth/Oauth2CallbackServlet.java]: I'm fairly certain this is a bad idea -- PII?

File: bookshelf/src/main/java/com/example/managedvms/gettingstartedjava/auth/LogoutServlet.java [hunk 4]
```diff
@@ -29,14 +34,35 @@
...
+      if (cookieMap.containsKey("token")) {
+        cookieMap.get("token").setMaxAge(0);
+        resp.addCookie(cookieMap.get("token"));
+        cookieMap.get("userEmail").setMaxAge(0);
+        resp.addCookie(cookieMap.get("userEmail"));
+        cookieMap.get("userImageUrl").setMaxAge(0);
+        resp.addCookie(cookieMap.get("userImageUrl"));
+        cookieMap.get("userId").setMaxAge(0);
+        resp.addCookie(cookieMap.get("userId"));
+      }
```
Comment (lesv) [inline bookshelf/src/main/java/com/example/managedvms/gettingstartedjava/auth/LogoutServlet.java]: PII?  Really?

**Comment by lesv** [comment]
Did the prior method w/ SessionStore work?  In a multi-instance environment?  If so, I'm inclined to keep the way you did it.  I think we should GVC in the AM to chat about this, 'caus maybe it's just late and I'm not thinking clearly.

**Comment by shun-fan** [comment]
Switching to datastore

**Pull request closed without merging**